### PR TITLE
[FIRRTL][Annotation] Use FieldID instead of strings to reference sub-target

### PR DESCRIPTION
**Pull request by hanchenye**
Currently, we are using an array of string to reference sub-target in annotations:
```
%bar = firrtl.reg %clock {annotations = [{target = ["[0]", ".baz"], one}]} : 
    (!firrtl.clock) -> !firrtl.vector<bundle<baz: vector<uint<1>, 2>>, 2>
```
In this patch, we replace the strings with `FieldID` introduced in #1045. The annotations above will look like:
```
%bar = firrtl.reg %clock {annotations = [{fieldID = 2 : i64, fieldIDRange = 2 : i64, one}]} : 
    (!firrtl.clock) -> !firrtl.vector<bundle<baz: vector<uint<1>, 2>>, 2>
```
Note that when the sub-target points to aggregate types (in the example above, `vector<uint<1>, 2>`), we introduce a `fieldIDRange` attribute to indicate the range of IDs of the aggregate type. Such that in the `-lower-types`, as long as a flattened field's ID is in the range of `[fieldID, fieldID + fieldIDRange]`, it should be annotated.

Meanwhile, this patch also adds the second verification mentioned in #1013.

### CODE REVIEW

**Comment by hanchenye** [comment]
One possible improvement here is to introduce a new struct attribute, such as:
```
{annotations = firrtl.subanno<[2,4], {one}>, ...}
```
But before that I'm curious whether people think the overall idea makes sense.

File: test/Dialect/FIRRTL/annotations.fir [hunk 1]
```diff
@@ -220,24 +220,24 @@ circuit Foo: %[[{"a":"{\"b\":null}"}]]
 ; // -----
 
 ; COM: Subfield/Subindex annotations should be parsed correctly on wires
-circuit Foo: %[[{"one":null,"target":"~Foo|Foo>bar.qux"},{"two":null,"target":"~Foo|Foo>bar[1].baz"} ]]
```
Comment (prithayan) [inline test/Dialect/FIRRTL/annotations.fir:223]: Is `one` not legal ? Seems like we can't annotate on `qux` without indexing into bar.

Comment (hanchenye) [reply test/Dialect/FIRRTL/annotations.fir:223]: Yea, this is illegal according to @seldridge. The current `-lower-types` will ignore this annotation in the lowering. With this patch we'll emit an error on this.

Comment (prithayan) [reply test/Dialect/FIRRTL/annotations.fir:223]: Right, that makes sense, Thanks

**Comment by prithayan** [comment]
Looks good to me.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp [hunk 2]
```diff
@@ -126,38 +126,35 @@ static void filterAnnotations(ArrayAttr annotations,
...
-      // If it ends with ']', then just remove it.
-      if (fNameStr.back() == ']')
-        fNameStr.erase(fNameStr.size() - 1);
-
-      targetStr += fNameStr;
-    }
-    // If no subfield attribute, then copy the annotation.
-    if (targetStr.empty()) {
-      loweredAttrs.push_back(opAttr);
+    auto fieldID = fieldIDAttr.cast<IntegerAttr>().getValue().getSExtValue();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:135]: Is this cast guaranteed to succeed?

File: test/Dialect/FIRRTL/annotations.fir [hunk 3]
```diff
@@ -220,24 +220,24 @@ circuit Foo: %[[{"a":"{\"b\":null}"}]]
...
+circuit Foo: %[[{"one":null,"target":"~Foo|Foo>bar[0]"},{"two":null,"target":"~Foo|Foo>bar[1].baz"} ]]
   module Foo:
     wire bar: {baz: UInt<1>, qux: UInt<1>}[2]
 
     ; CHECK-LABEL: module {
-    ; CHECK: %bar = firrtl.wire  {annotations = [{one, target = [".qux"]}, {target = ["[1]", ".baz"], two}]} : !firrtl.vector<bundle<baz: uint<1>, qux: uint<1>>, 2>
+    ; CHECK: %bar = firrtl.wire  {annotations = [{fieldID = 1 : i64, fieldIDRange = 2 : i64, one}, {fieldID = 5 : i64, two}]} : !firrtl.vector<bundle<baz: uint<1>, qux: uint<1>>, 2>
```
Comment (seldridge) [inline test/Dialect/FIRRTL/annotations.fir:228]: I really like this representation.

File: test/Dialect/FIRRTL/annotations-errors.fir [hunk 4]
```diff
@@ -73,3 +73,22 @@ circuit Foo: %[[{"target":""}]]
...
+
+; // -----
+
+; COM: Subfield/Subindex annotations should be parsed correctly on registers
+
+circuit Foo: %[[{"one":null,"target":"~Foo|Foo>bar.qux"},{"two":null,"target":"~Foo|Foo>bar[1].baz"} ]]
+  module Foo:
+    input clock: Clock
+; expected-error @+1 {{invalid sub-index name .qux}}
+    reg bar: {baz: UInt<1>, qux: UInt<1>}[2], clock
```
Comment (seldridge) [inline test/Dialect/FIRRTL/annotations-errors.fir:94]: There are a lot of error paths that aren't covered here. Can you add those?

**Review by seldridge — COMMENTED** [review]
This is looking great!

Some minor comments throughout. I'd like to take one more look after this gets updated.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 5]
```diff
@@ -951,23 +1008,66 @@ void FIRParser::getAnnotations(Twine target, ArrayAttr &annotations) {
...
+      continue;
+    }
+
+    // Get field ID from the subscripts.
+    if (!type)
+      continue;
+    ArrayAttr subscripts = targetAttr.cast<ArrayAttr>();
+    auto fieldID = getFieldIDFromSubscripts(loc, subscripts, type);
+    if (!fieldID)
+      continue;
```
[the remark this answers is not in the document]
Comment (hanchenye) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:1033]: Right! I was thinking about this. Will file a PR after this one.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 6]
```diff
@@ -928,21 +933,73 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
...
+  for (auto subscript : subscripts) {
+    auto nameStr = subscript.cast<StringAttr>().getValue();
+    // Unflip if the current type is flipped type.
+    if (auto flipType = currentType.dyn_cast<FlipType>())
+      currentType = flipType.getElementType();
+
+    if (auto bundleType = currentType.dyn_cast<BundleType>()) {
+      // Target should begin with ".".
+      if (nameStr.front() != '.')
+        return emitError(loc, "invalid sub-field name " + nameStr), None;
```
[the remark this answers is not in the document]
Comment (hanchenye) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:954]: In this case, we have not built the operation when constructing the annotations. For now, it's generating the errors at the beginning of the corresponding operation, which should look good.

**Comment by hanchenye** [comment]
@seldridge Several updates:
1. Error messages are improved and all error paths are tested;
2. Now `FIRAnnotation` is generating array of `IntegerAttr` and `StringAttr`, simplifying the logic in `FIRParser`;
3. Introduce a new attribute `SubAnnotationAttr` representing as`#firrtl.subAnno<fieldID = [2,4], {one}>`, which makes the code in `LowerTypes` super clean.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 7]
```diff
@@ -12,6 +12,7 @@
 
 #include "circt/Dialect/FIRRTL/FIRRTLOps.h"
 #include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAttributes.h"
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:15]: Why is this necessary? (I would've thought to see some change to this file...)

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp [hunk 8]
```diff
@@ -43,17 +43,29 @@ static StringRef parseSubFieldSubIndexAnnotations(StringRef target,
...
+    if (c == ']') {
+      // Create a IntegerAttr with the previous sub-index token.
+      APInt subIndex;
+      if (!temp.str().getAsInteger(10, subIndex))
+        annotationVec.push_back(IntegerAttr::get(IntegerType::get(context, 64),
+                                                 subIndex.getSExtValue()));
+      else
+        // We don't have a good way to emit error here. This will be reported as
+        // an error in the FIRRTL parser.
+        annotationVec.push_back(StringAttr::get(context, temp));
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:56]: Bad architecture on my part. 😅 We should figure out how to do this better.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 9]
```diff
@@ -339,12 +339,17 @@ struct FIRParser {
   /// Add annotations from the source manager, if an annotation file was added.
   ParseResult importAnnotationFile(SMLoc loc);
 
-  /// Populate a vector of annotations for a given Target.  If the annotations
-  /// parameter is non-empty, then this will be appended to.
-  void getAnnotations(Twine target, ArrayAttr &annotations);
+  Optional<std::pair<unsigned, unsigned>>
+  getFieldIDFromTokens(SMLoc loc, ArrayAttr tokens, Type type);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:343]: Can you include some documentation about this function?

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 10]
```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
   return result;
 }
 
+Optional<std::pair<unsigned, unsigned>>
+FIRParser::getFieldIDFromTokens(SMLoc loc, ArrayAttr tokens, Type type) {
+  if (!type)
+    return None;
+  if (tokens.empty())
+    return std::pair<unsigned, unsigned>(0, 0);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:949]: Consider using:
```suggestion
    return {{0, 0}};
```

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 11]
```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
...
+
+  auto currentType = type.cast<FIRRTLType>();
+  unsigned id = 0;
+  unsigned idRange = 0;
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:953]: Nit:
```suggestion
  unsigned id = 0, idRange = 0;
```

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 12]
```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
...
+      tokensStr.append(strAttr.getValue());
+    tokensStr.push_back('_');
+  }
+  tokensStr.pop_back();
+
+  unsigned tokenIdx = 1;
+  for (auto token : tokens) {
+    // Unflip if the current type is flipped type.
+    if (auto flipType = currentType.dyn_cast<FlipType>())
+      currentType = flipType.getElementType();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:970]: This should be able to be removed after a rebase (since Flip is now gone. 😄 ).

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 13]
```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
...
+
+    } else {
+      emitError(loc, "the " + std::to_string(tokenIdx) + "-th token of " +
+                         tokensStr + " expects an aggregate type");
+      return None;
+    }
+    ++tokenIdx;
+  }
+
+  return std::pair<unsigned, unsigned>(id, id + idRange);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:1021]: Consider:
```suggestion
  return {{id, id + idRange}};
```

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 14]
```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
...
+        emitError(loc, "the " + std::to_string(tokenIdx) + "-th token of " +
+                           tokensStr + " is not found in the bundle");
+        return None;
+      }
+
+      id += bundleType.getFieldID(index.getValue());
+      currentType = bundleType.getElementType(subField);
+      idRange = currentType.getMaxFieldID();
+
+    } else if (auto vectorType = currentType.dyn_cast<FVectorType>()) {
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:993]: You can remove the long if/else if/else chain if you move initialize `tokenIdx` to zero, increment it at the start of the loop, and then exit via `continue` in each of the cases. This would produce something possibly more readable of:
```c++
if (foo) {
  // ... 
  continue;
}
if (bar) {
  // ...
  continue;
}
```

Alternatively, this could make use of a three-case `TypeSwitch`.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp [hunk 15]
```diff
@@ -116,65 +117,29 @@ static FIRRTLType getCanonicalAggregateType(Type originalType) {
...
     return;
 
   for (auto opAttr : annotations) {
-    auto di = opAttr.dyn_cast<DictionaryAttr>();
-    if (!di) {
+    if (auto subAnno = opAttr.dyn_cast<SubAnnotationAttr>()) {
+      if (targetFieldID >= subAnno.getMinFieldID() &&
+          targetFieldID <= subAnno.getMaxFieldID())
+        loweredAttrs.push_back(subAnno.getAnnotations());
+    } else
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:129]: Amazing linecount reduction. :100:

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 16]
```diff
@@ -936,21 +941,96 @@ ParseResult FIRParser::importAnnotationFile(SMLoc loc) {
...
+  // Construct a string for error emission.
+  SmallString<16> tokensStr;
+  for (auto token : tokens) {
+    if (auto intAttr = token.dyn_cast<IntegerAttr>())
+      intAttr.getValue().toStringSigned(tokensStr);
+    else if (auto strAttr = token.dyn_cast<StringAttr>())
+      tokensStr.append(strAttr.getValue());
+    tokensStr.push_back('_');
+  }
+  tokensStr.pop_back();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:964]: Can this not flatten the tokens and instead use `[`, `]`, and `.`? The error message could then be `foo.bar[0]` instead of `foo_bar_0`.

**Review by seldridge — COMMENTED** [review]
The use of field ids here is awesome. Some nitty comments throughout.

This also needs a rebase. :smile:

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 7 above)
Comment (hanchenye) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:15]: This is because `FIRRTL.cpp.inc` are using the new attribute we introduced :eyes:

**Comment by hanchenye** [comment]
The rebase has been completed. This PR should be ready for review. The `getAnnotations` related methods are updated a lot in order to support the port annotation feature.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 7 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:15]: Got it. That makes sense and this is then the minimal location where it should be put. :100:

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 17]
```diff
@@ -2327,68 +2599,26 @@ ParseResult FIRStmtParser::parseInstance() {
...
-          dictAttrVec.push_back(builder.getNamedAttr("target", newTargetAttr));
-        }
-      } else
-        dictAttrVec.push_back(namedAttr);
-    }
+  auto annotations =
+      getSplittedAnnotations({getModuleTarget() + ">" + id,
+                              getModuleTarget() + "/" + id + ":" + moduleName},
+                             startTok.getLoc(), resultNamesAndTypes);
+  auto name = hasDontTouch(annotations.first) ? id : filterUselessName(id);
```
[the remark this answers is not in the document]
Comment (hanchenye) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:2618]: Good point, will fix this issue!

File: include/circt/Dialect/FIRRTL/FIRRTLTypes.td [hunk 18]
```diff
@@ -105,8 +105,30 @@ def MemDirAttr: I32EnumAttr<"MemDirAttr", "Memory Direction Enum",
...
+  : AttrDef<FIRRTLDialect, "SubAnnotation", [], "::mlir::Attribute"> {
+  let summary = "An Annotation that targets part of what it's attached to";
+  let description = [{
+    An Annotation that is only applicable to part of what it is attached to.
+    This uses a range of field IDs to indicate to which fields it is applicable.
+  }];
+  let mnemonic = "subAnno";
+  let parameters = (ins "int64_t":$minFieldID, "int64_t":$maxFieldID,
+                        "DictionaryAttr":$annotations);
+}
```
Comment (prithayan) [inline include/circt/Dialect/FIRRTL/FIRRTLTypes.td:131]: Nice, `SubAnnotationAttr` is a really neat idea.

**Comment by lattner** [comment]
cool!
